sqlx-postgres/src/types/ts_query.rs

```diff
@@ -0,0 +1,213 @@
+                    let mut text = String::new().into_bytes();
+
+                    reader.read_until(b'\0', &mut text)?;
+
+                    let text = String::from_utf8(text)?;
+                    let prefix = reader.read_u8()?;
+
+                    entries.push(Entry::Value(Value {
+                        weight,
+                        text,
+                        prefix,
+                    }));
+                }
+                EntryType::Operator => {
+                    let operator: Operators = reader.read_u8()?.try_into()?;
+                    let distance = if let Operators::Phrase = operator {
+                        Some(reader.read_u16::<BigEndian>()?)
+                    } else {
+                        None
+                    };
+
+                    entries.push(Entry::Operator(Operator { operator, distance }));
+                }
+            }
+        }
+
+        Ok(TsQuery { entries })
+    }
+}
+
+impl TryInto<Vec<u8>> for &TsQuery {
+    type Error = BoxDynError;
+
+    fn try_into(self) -> Result<Vec<u8>, Self::Error> {
+        let buf: &mut Vec<u8> = &mut vec![];
+
+        buf.write_u32::<BigEndian>(u32::try_from(self.entries.len())?)?;
+
+        for entry in &self.entries {
+            match entry {
+                Entry::Operator(operator) => {
+                    buf.write_u8(EntryType::Operator.into())?;
+                    buf.write_u8(operator.operator.into())?;
+
+                    if let Some(distance) = operator.distance {
+                        buf.write_u16::<BigEndian>(distance)?;
+                    }
+                }
+                Entry::Value(value) => {
+                    buf.write_u8(EntryType::Value.into())?;
+                    buf.write_u8(value.weight)?;
+
+                    buf.write(value.text.as_bytes())?;
+                    buf.write(&[b'\0'])?;
+
+                    buf.write_u8(value.prefix)?;
+                }
+            }
+        }
+
+        buf.flush()?;
+
+        Ok(buf.to_vec())
+    }
+}
+
+impl Type<Postgres> for TsQuery {
+    fn type_info() -> PgTypeInfo {
+        PgTypeInfo::TS_QUERY
+    }
+}
+
+impl PgHasArrayType for TsQuery {
+    fn array_type_info() -> PgTypeInfo {
+        PgTypeInfo::TS_QUERY_ARRAY
+    }
+}
+
+impl Encode<'_, Postgres> for TsQuery {
+    fn encode_by_ref(&self, buf: &mut PgArgumentBuffer) -> IsNull {
+        if let Ok(encoded_ts_query) = <&TsQuery as TryInto<Vec<u8>>>::try_into(self) {
+            buf.extend_from_slice(encoded_ts_query.as_slice());
+
+            IsNull::No
+        } else {
+            IsNull::Yes
+        }
+    }
+}
+
+impl Decode<'_, Postgres> for TsQuery {
+    fn decode(value: PgValueRef<'_>) -> Result<Self, BoxDynError> {
+        match value.format() {
+            PgValueFormat::Binary => {
+                let bytes = value.as_bytes()?;
+                let ts_query = bytes.try_into()?;
+
+                Ok(ts_query)
+            }
+            _ => unimplemented!(),
```

Comment:
Can return a `Err(...unimplemented)`? That way we can easily catch it. (Just a suggestion)

Reply:
I didn't review `ts_query.rs`, because I know nothing about `TsQuery`.

Reply:
@wyhaya Got it. Thanks. I'll take another look at TsQuery based on the documentation here: https://doxygen.postgresql.org/ts__type_8h_source.html.

Reply:
@wyhaya Turns out, that TSQuery implementation is nowhere near ready. I'm compelled to take it out of this PR and do it separately so that we don't block TSVector. Since they can work independent to each other, I will do so.